node/lib/DeviceInterface.ts

```diff
@@ -0,0 +1,122 @@
+     */
+    public readonly name: string;
+
+    /**
+     * Optional English description of the interface, to be used as developer documentation.
+     */
+    public readonly description?: string;
+}
+
+/**
+ * Specifies an interface that can be implemented by a device.
+ *
+ * Interfaces are composable, meaning they can reference other interfaces, which reference
+ * other interfaces, etc. So getting a complete view of all of an interface's properties
+ * and methods requires resolving and following all references. (Duplicate interface
+ * references are ignored; cycles are not allowed.) Interface resolution is provided
+ * by another class (TBD).
+ *
+ * Property and method names are allowed to conflict across different interfaces, even
+ * ones that implemented by the same device. When accessing a device, an interface
+ * context is always specifid.
+ */
+export class DeviceInterface extends DeviceCharacteristic {
+    /**
+     * List of properties declared by this interface.
+     */
+    public readonly properties: DeviceProperty[];
+
+    /**
+     * List of methods declared by this interface.
+     */
+    public readonly methods: DeviceMethod[];
+
+    /**
+     * List of referenced interfaces declared by this interface. All properties and
+     * methods from directly and indirectly referenced interfaces are effectively
+     * included in this interface. (It's conceptually similar to interface inheritance.)
+     */
+    public readonly references: DeviceInterface[];
+}
+
+/**
+ * A property declared by a DeviceInterface. Properties can be readable,
+ * writeable, and/or notifiable. (A property that is notify-only is equivalent
+ * to an AllJoyn signal.)
+ */
+export class DeviceProperty extends DeviceCharacteristic {
+    /**
+     * Name of the interface that declared this property.
+     */
+    public readonly interfaceName: string;
+
+    /**
+     * Whether the property is readable.
+     */
+    public readonly canRead: boolean;
+
+    /**
+     * Whether the property is writeable.
+     */
+    public readonly canWrite: boolean;
+
+    /**
+     * Whether the property supports notifications.
+     *
+     * Notifications can be subscribed to using the standard node events API.
+     *
+     * For a property that is also readable, a notification typically represents
+     * a property-changed event. For a notify-only property, the meaning of the
+     * notification may be different. Either way, the type of the notification
+     * event object conforms to the property type schema.
+     */
+    public readonly canNotify: boolean;
+
+    /**
+     * JSON schema that specifies the type of the property.
+     */
+    public readonly propertyType: Schema;
+}
+
+/**
+ * A method declared by a device interface.
+ */
+export class DeviceMethod extends DeviceCharacteristic {
+    /**
+     * Name of the interface that declared this method.
+     */
+    public readonly interfaceName: string;
```

Comment:
I understand why it's not obvious yet. I'll add more comments.

The purpose of this will be clearer when we have a way to resolve referenced interfaces and then get ALL the properties and methods on an interface, including those from references. Then you will need to know the name of the interface that actually declared the method in order to call the method properly, because there could be multiple methods with the same name declared by different referenced interfaces.

---
In reply to: [71903851](https://github.com/opent2t/opent2t/pull/1#discussion_r71903851) [](ancestors = 71903851)